### PR TITLE
Use Vue to refactor pull merge UI

### PR DESCRIPTION
**Pull request by wxiaoguang**
This PR would help

* https://github.com/go-gitea/gitea/pull/19648
* https://github.com/go-gitea/gitea/pull/9307

to get a clear code base.

And this PR use real code to show how to rewrite UI to Vue

At the moment, to keep the UI the same, some Fomantic UI styles are still being used.

I just did some simple tests, it should work as expected. Feel free to report bugs or edit on this PR directly.

### CODE REVIEW

**Comment by 6543** [comment]
Screenshot at 2022-05-07 22-01-01
Screenshot at 2022-05-07 22-01-59

**Comment by wxiaoguang** [comment]
@6543 I am cleaning comments, the eslint and script-module cann't work together. pull & try again.

**Review by 6543 — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

File: templates/repo/issue/view_content/pull.tmpl [hunk 1]
```diff
@@ -322,176 +322,62 @@
...
+										{
+											'name': 'rebase',
+											'allowed': {{$prUnit.PullRequestsConfig.AllowRebase}},
+											'textDoMerge': {{$.i18n.Tr "repo.pulls.rebase_merge_pull_request"}},
+											'hideMergeMessage': true,
+										},
+										{
+											'name': 'rebase-merge',
+											'allowed': {{$prUnit.PullRequestsConfig.AllowRebaseMerge}},
+											'textDoMerge': {{$.i18n.Tr "repo.pulls.rebase_merge_commit_pull_request"}},
```
Comment (delvh) [inline templates/repo/issue/view_content/pull.tmpl:362]: Has `rebase-merge` a merge message?

File: templates/repo/issue/view_content/pull.tmpl [hunk 2]
```diff
@@ -322,176 +322,62 @@
...
-												{{end}}
-											</div>
-										</div>
-									{{end}}
-								</div>
-							</div>
+
+							<script>
+								<!-- /* eslint-disable */ -->
+								(() => {
```
Comment (delvh) [inline templates/repo/issue/view_content/pull.tmpl:330]: What's the benefit of wrapping it inside an immediately called function here?

File: templates/repo/issue/view_content/pull.tmpl [hunk 3]
```diff
@@ -322,176 +322,62 @@
...
+
+										'allOverridableChecksOk': {{not $notAllOverridableChecksOk}},
+										'pullHeadCommitID': {{.PullHeadCommitID}},
+										'mergeTitleFieldText': {{.Issue.PullRequest.GetDefaultMergeMessage $.Context}},
+										'mergeMessageFieldPlaceHolder': {{$.i18n.Tr "repo.editor.commit_message_desc"}},
+										'mergeMessageFieldText': 'Reviewed-on: ' + {{$.Issue.HTMLURL}} + '\n' + {{$approvers}},
+										'isPullBranchDeletable': {{.IsPullBranchDeletable}},
+										'defaultDeleteBranchAfterMerge': {{$prUnit.PullRequestsConfig.DefaultDeleteBranchAfterMerge}},
+									};
+									mergeForm['mergeStyles'] = [
```
Comment (delvh) [inline templates/repo/issue/view_content/pull.tmpl:345]: I think you could also declare it directly inside the object declaration.

File: templates/repo/issue/view_content/pull.tmpl [hunk 4]
```diff
@@ -322,176 +322,62 @@
...
+									const mergeForm = {
+										'baseLink': {{.Link}},
+										'textCancel': {{$.i18n.Tr "cancel"}},
+										'textDeleteBranch': {{$.i18n.Tr "repo.branch.delete" .HeadTarget}},
+
+										'allOverridableChecksOk': {{not $notAllOverridableChecksOk}},
+										'pullHeadCommitID': {{.PullHeadCommitID}},
+										'mergeTitleFieldText': {{.Issue.PullRequest.GetDefaultMergeMessage $.Context}},
+										'mergeMessageFieldPlaceHolder': {{$.i18n.Tr "repo.editor.commit_message_desc"}},
+										'mergeMessageFieldText': 'Reviewed-on: ' + {{$.Issue.HTMLURL}} + '\n' + {{$approvers}},
```
Comment (delvh) [inline templates/repo/issue/view_content/pull.tmpl]: I prefer the shorter versions:
```suggestion
										'mergeCommitTitle': {{.Issue.PullRequest.GetDefaultMergeMessage $.Context}},
										'mergeMessagePlaceholder': {{$.i18n.Tr "repo.editor.commit_message_desc"}},
										'mergeMessage': 'Reviewed-on: ' + {{$.Issue.HTMLURL}} + '\n' + {{$approvers}},
```
What about you?

File: web_src/js/components/PullRequestMergeForm.vue [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+    <template v-if="!showActionForm">
+      <div class="ui buttons merge-button" :class="[mergeForm.allOverridableChecksOk?'green':'red']" @click="showActionForm=true">
+        <button class="ui button">
+          <svg-icon name="octicon-git-merge"/>
+          <span class="button-text">{{ mergeStyleDetail.textDoMerge }}</span>
+        </button>
+        <div class="ui dropdown icon button no-text" @click.stop="showMergeStyleMenu = !showMergeStyleMenu" v-if="mergeStyleAllowedCount>1">
+          <svg-icon name="octicon-triangle-down" :size="14"/>
+          <div class="menu" :class="{'show':showMergeStyleMenu}">
```
Comment (delvh) [inline web_src/js/components/PullRequestMergeForm.vue:40]: What does `:class="{'show':showMergeStyleMenu}"` do?

File: templates/repo/issue/view_content/pull.tmpl [hunk 6]
```diff
@@ -322,176 +322,62 @@
...
-												<div class="item{{if eq .MergeStyle "manually-merged"}} active selected{{end}}" data-do="manually-merged">{{$.i18n.Tr "repo.pulls.merge_manually"}}</div>
-												{{end}}
-											</div>
-										</div>
-									{{end}}
-								</div>
-							</div>
+
+							<script>
+								<!-- /* eslint-disable */ -->
```
Comment (wxiaoguang) [inline templates/repo/issue/view_content/pull.tmpl:329]: For future readers: at the moment, "eslint-disable" + "script(type=module)" + go-template can not work together.

**Comment by 6543** [comment]
@wxiaoguang please resolve config

**Comment by wxiaoguang** [comment]
Need to confirm a change first: https://github.com/go-gitea/gitea/pull/18177/files#r867490307

**Comment by wxiaoguang** [comment]
I think that the change in pull.tmpl on main branch might be wrong. So I just keep the detailed message textarea for squash merge in this PR. Updated.

**Comment by duhanmin** [comment]
@wxiaoguang Merge conflict resolution?

**Review by delvh — APPROVED** [review]

**Pull request merged**